File: sdk/typescript/src/adapters/slack.ts
/**
 * Slack adapter — send Bernstein task notifications via Incoming Webhooks or Web API.
 *
 * Uses the built-in `fetch` API (Node.js 18+). No extra dependencies.
 *
 * @example
 * ```ts
 * import { SlackAdapter } from '@bernstein/sdk/adapters/slack';
 *
 * const slack = new SlackAdapter({ webhookUrl: process.env.SLACK_WEBHOOK_URL });
 * await slack.notifyTaskCompleted('abc123', 'Fix login regression', 'backend', 'Patched null-check');
 * ```
 */

import type { TaskStatus } from '../models.js';

export interface SlackAdapterOptions {
  webhookUrl?: string;
  botToken?: string;
  channel?: string;
  /** Slack user/group to @mention on failures (e.g. "<!here>" or "<@U123>"). */
  mentionOnFailure?: string;
}

export interface SlackBlock {
  type: string;
  [key: string]: unknown;
}

export class SlackAdapter {
  private readonly webhookUrl: string;
  private readonly botToken: string;
  private readonly channel: string;
  private readonly mentionOnFailure: string;

  constructor(options: SlackAdapterOptions = {}) {
    this.webhookUrl = options.webhookUrl ?? '';
    this.botToken = options.botToken ?? '';
    this.channel = options.channel ?? '';
    this.mentionOnFailure = options.mentionOnFailure ?? '';

    if (!this.webhookUrl && !this.botToken) {
      console.warn('SlackAdapter: no webhook URL or bot token configured');
    }
  }

  static fromEnv(): SlackAdapter {
    return new SlackAdapter({
      webhookUrl: process.env['SLACK_WEBHOOK_URL'] ?? '',
      botToken: process.env['SLACK_BOT_TOKEN'] ?? '',
      channel: process.env['SLACK_CHANNEL'] ?? '',
      mentionOnFailure: process.env['SLACK_MENTION_ON_FAILURE'] ?? '',
    });
  }

  async notifyTaskCompleted(
    taskId: string,
    title: string,
    role: string,
    resultSummary = '',
    channel?: string,
  ): Promise<void> {
    const blocks = taskCompletedBlocks(taskId, title, role, resultSummary);
    await this.post(blocks, channel ?? this.channel);
  }

  async notifyTaskFailed(
    taskId: string,
    title: string,
    role: string,
    error = '',
    channel?: string,
  ): Promise<void> {
    const blocks = taskFailedBlocks(taskId, title, role, error, this.mentionOnFailure);
    await this.post(blocks, channel ?? this.channel);
  }

  async notifyTaskCreated(
    taskId: string,
    title: string,
    role: string,
    priority = 2,
    channel?: string,
  ): Promise<void> {
    const blocks = taskCreatedBlocks(taskId, title, role, priority);
    await this.post(blocks, channel ?? this.channel);
  }

  async postMessage(
    text: string,
    blocks?: SlackBlock[],
    channel?: string,
  ): Promise<void> {
    await this.post(blocks ?? [], channel ?? this.channel, text);
  }

  // -------------------------------------------------------------------------
  // Internal
  // -------------------------------------------------------------------------

  private async post(
    blocks: SlackBlock[],
    channel: string,
    text?: string,
  ): Promise<void> {
    const fallbackText = text ?? blocksToText(blocks);
    try {
      if (this.webhookUrl) {
        await fetchJson(this.webhookUrl, { blocks, text: fallbackText });
      } else if (this.botToken && channel) {
        await fetchJson(
          'https://slack.com/api/chat.postMessage',
          { channel, blocks, text: fallbackText },
          { Authorization: `Bearer ${this.botToken}` },
        );
      }
    } catch (err) {
      console.warn('SlackAdapter: failed to post message:', err);
    }
  }
}

// ---------------------------------------------------------------------------
// Block Kit builders
// ---------------------------------------------------------------------------

function taskCompletedBlocks(
  taskId: string,
  title: string,
  role: string,
  summary: string,
): SlackBlock[] {
  const fields: Record<string, unknown>[] = [
    { type: 'mrkdwn', text: `*Title:*\n${title}` },
    { type: 'mrkdwn', text: `*Role:*\n\`${role}\`` },
  ];
  if (summary) {
    fields.push({ type: 'mrkdwn', text: `*Result:*\n${summary.slice(0, 300)}` });
  }
  return [
    { type: 'section', text: { type: 'mrkdwn', text: `:white_check_mark: *Task completed* — \`${taskId}\`` } },
    { type: 'section', fields },
  ];
}

function taskFailedBlocks(
  taskId: string,
  title: string,
  role: string,
  error: string,
  mention: string,
): SlackBlock[] {
  let header = `:x: *Task failed* — \`${taskId}\``;
  if (mention) header = `${mention} ${header}`;
  const blocks: SlackBlock[] = [
    { type: 'section', text: { type: 'mrkdwn', text: header } },
    {
      type: 'section',
      fields: [
        { type: 'mrkdwn', text: `*Title:*\n${title}` },
        { type: 'mrkdwn', text: `*Role:*\n\`${role}\`` },
      ],
    },
  ];
  if (error) {
    blocks.push({
      type: 'section',
      text: { type: 'mrkdwn', text: `*Error:*\n\`\`\`${error.slice(0, 500)}\`\`\`` },
    });
  }
  return blocks;
}

function taskCreatedBlocks(
  taskId: string,
  title: string,
  role: string,
  priority: number,
): SlackBlock[] {
  const emoji =
    priority === 1 ? ':rotating_light:' : priority === 3 ? ':white_circle:' : ':blue_circle:';
  return [
    {
      type: 'section',
      text: { type: 'mrkdwn', text: `${emoji} *New task* — \`${taskId}\`` },
      fields: [
        { type: 'mrkdwn', text: `*Title:*\n${title}` },
        { type: 'mrkdwn', text: `*Role:*\n\`${role}\`` },
      ],
    },
  ];
}

function blocksToText(blocks: SlackBlock[]): string {
  for (const block of blocks) {
    const textObj = block['text'];
    if (textObj && typeof textObj === 'object') {
      const t = (textObj as Record<string, unknown>)['text'];
      if (typeof t === 'string' && t) return t;
    }
  }
  return 'Bernstein notification';
}

async function fetchJson(
  url: string,
  body: unknown,
  extraHeaders: Record<string, string> = {},
): Promise<void> {
  const res = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', ...extraHeaders },
    body: JSON.stringify(body),
  });
  if (!res.ok) {
    const text = await res.text().catch(() => '');
    throw new Error(`HTTP ${res.status}: ${text.slice(0, 200)}`);
  }
}

// Re-export TaskStatus so callers can build notify calls from task data
export type { TaskStatus };

File: sdk/python/src/bernstein_sdk/adapters/slack.py
"""Slack adapter — send Bernstein task notifications to Slack channels.

Sends Block Kit messages to Slack via Incoming Webhooks or the Web API.

Setup
-----
Option A — Incoming Webhook (simplest)::

    export SLACK_WEBHOOK_URL=https://hooks.slack.com/services/T.../B.../...

Option B — Bot token (supports specifying channels dynamically)::

    export SLACK_BOT_TOKEN=xoxb-...

Usage::

    from bernstein_sdk.adapters.slack import SlackAdapter, TaskEvent

    adapter = SlackAdapter.from_env()
    adapter.notify_task_completed(
        task_id="abc123",
        title="Fix login regression",
        role="backend",
        result_summary="Patched null-check in auth.py",
    )
"""

from __future__ import annotations

import json
import logging
import os
import threading
import urllib.request
from typing import Any

log = logging.getLogger(__name__)

_SLACK_API_BASE = "https://slack.com/api"


class SlackAdapter:
    """Send Bernstein task events to Slack.

    Supports both Incoming Webhooks (simpler) and the Slack Web API (more
    control).  All HTTP calls are dispatched on daemon threads so they never
    block the caller.

    Args:
        webhook_url: Slack Incoming Webhook URL.  If set, notifications are
            sent here regardless of *bot_token* / *channel*.
        bot_token: Slack bot OAuth token (``xoxb-...``).  Required if
            *webhook_url* is not set.
        channel: Default Slack channel ID or name (e.g. ``"#dev-notifications"``).
        mention_on_failure: Slack user/group ID to ``@mention`` when a task
            fails (e.g. ``"<!here>"`` or ``"<@U12345>"``).
    """

    def __init__(
        self,
        webhook_url: str = "",
        bot_token: str = "",
        channel: str = "",
        mention_on_failure: str = "",
    ) -> None:
        self._webhook_url = webhook_url
        self._bot_token = bot_token
        self._channel = channel
        self._mention_on_failure = mention_on_failure

    @classmethod
    def from_env(cls) -> SlackAdapter:
        """Construct from environment variables.

        Reads ``SLACK_WEBHOOK_URL`` and/or ``SLACK_BOT_TOKEN`` /
        ``SLACK_CHANNEL``.
        """
        return cls(
            webhook_url=os.getenv("SLACK_WEBHOOK_URL", ""),
            bot_token=os.getenv("SLACK_BOT_TOKEN", ""),
            channel=os.getenv("SLACK_CHANNEL", ""),
            mention_on_failure=os.getenv("SLACK_MENTION_ON_FAILURE", ""),
        )

    # ------------------------------------------------------------------
    # Public notification helpers
    # ------------------------------------------------------------------

    def notify_task_completed(
        self,
        task_id: str,
        title: str,
        role: str,
        result_summary: str = "",
        channel: str = "",
    ) -> None:
        """Post a task-completed notification."""
        blocks = _task_completed_blocks(
            task_id=task_id, title=title, role=role, summary=result_summary
        )
        self._post_async(blocks=blocks, channel=channel or self._channel)

    def notify_task_failed(
        self,
        task_id: str,
        title: str,
        role: str,
        error: str = "",
        channel: str = "",
    ) -> None:
        """Post a task-failed notification (optionally @mentioning a user)."""
        mention = self._mention_on_failure
        blocks = _task_failed_blocks(
            task_id=task_id, title=title, role=role, error=error, mention=mention
        )
        self._post_async(blocks=blocks, channel=channel or self._channel)

    def notify_task_created(
        self,
        task_id: str,
        title: str,
        role: str,
        priority: int = 2,
        channel: str = "",
    ) -> None:
        """Post a task-created notification."""
        blocks = _task_created_blocks(
            task_id=task_id, title=title, role=role, priority=priority
        )
        self._post_async(blocks=blocks, channel=channel or self._channel)

    def post_message(
        self, text: str, blocks: list[dict[str, Any]] | None = None, channel: str = ""
    ) -> None:
        """Post an arbitrary text or Block Kit message."""
        self._post_async(
            text=text, blocks=blocks or [], channel=channel or self._channel
        )

    # ------------------------------------------------------------------
    # Internal
    # ------------------------------------------------------------------

    def _post_async(
        self,
        blocks: list[dict[str, Any]],
        text: str = "",
        channel: str = "",
    ) -> None:
        """Dispatch a Slack message on a daemon thread (fire-and-forget)."""
        if not self._webhook_url and not self._bot_token:
            log.warning("SlackAdapter: no webhook URL or bot token configured")
            return
        target_channel = channel or self._channel

        def _send() -> None:
            try:
                if self._webhook_url:
                    _post_webhook(
                        self._webhook_url,
                        blocks=blocks,
                        text=text or _blocks_fallback(blocks),
                    )
                elif self._bot_token and target_channel:
                    _post_web_api(
                        token=self._bot_token,
                        channel=target_channel,
                        blocks=blocks,
                        text=text or _blocks_fallback(blocks),
                    )
                else:
                    log.warning(
                        "SlackAdapter: no channel configured for bot token delivery"
                    )
            except Exception as exc:
                log.warning("SlackAdapter: failed to post message: %s", exc)

        t = threading.Thread(target=_send, daemon=True)
        t.start()


# ---------------------------------------------------------------------------
# Block Kit builders
# ---------------------------------------------------------------------------


def _task_completed_blocks(
    task_id: str, title: str, role: str, summary: str
) -> list[dict[str, Any]]:
    text = f":white_check_mark: *Task completed* — `{task_id}`"
    fields = [
        {"type": "mrkdwn", "text": f"*Title:*\n{title}"},
        {"type": "mrkdwn", "text": f"*Role:*\n`{role}`"},
    ]
    if summary:
        fields.append({"type": "mrkdwn", "text": f"*Result:*\n{summary[:300]}"})
    return [
        {"type": "section", "text": {"type": "mrkdwn", "text": text}},
        {"type": "section", "fields": fields},
    ]


def _task_failed_blocks(
    task_id: str, title: str, role: str, error: str, mention: str
) -> list[dict[str, Any]]:
    header = f":x: *Task failed* — `{task_id}`"
    if mention:
        header = f"{mention} {header}"
    fields = [
        {"type": "mrkdwn", "text": f"*Title:*\n{title}"},
        {"type": "mrkdwn", "text": f"*Role:*\n`{role}`"},
    ]
    blocks: list[dict[str, Any]] = [
        {"type": "section", "text": {"type": "mrkdwn", "text": header}},
        {"type": "section", "fields": fields},
    ]
    if error:
        blocks.append(
            {
                "type": "section",
                "text": {"type": "mrkdwn", "text": f"*Error:*\n```{error[:500]}```"},
            }
        )
    return blocks


def _task_created_blocks(
    task_id: str, title: str, role: str, priority: int
) -> list[dict[str, Any]]:
    priority_emoji = {
        1: ":rotating_light:",
        2: ":blue_circle:",
        3: ":white_circle:",
    }.get(priority, ":blue_circle:")
    return [
        {
            "type": "section",
            "text": {
                "type": "mrkdwn",
                "text": f"{priority_emoji} *New task* — `{task_id}`",
            },
            "fields": [
                {"type": "mrkdwn", "text": f"*Title:*\n{title}"},
                {"type": "mrkdwn", "text": f"*Role:*\n`{role}`"},
            ],
        }
    ]


def _blocks_fallback(blocks: list[dict[str, Any]]) -> str:
    """Extract a plain-text fallback from a Block Kit message."""
    for block in blocks:
        text_obj = block.get("text")
        if isinstance(text_obj, dict):
            text = text_obj.get("text", "")
            if text:
                return text
    return "Bernstein notification"


# ---------------------------------------------------------------------------
# HTTP helpers (no extra dependencies)
# ---------------------------------------------------------------------------


def _post_webhook(url: str, blocks: list[dict[str, Any]], text: str) -> None:
    payload = {"blocks": blocks, "text": text}
    data = json.dumps(payload).encode()
    req = urllib.request.Request(
        url, data=data, headers={"Content-Type": "application/json"}, method="POST"
    )
    with urllib.request.urlopen(req, timeout=5):
        pass  # Response body not needed


def _post_web_api(
    token: str, channel: str, blocks: list[dict[str, Any]], text: str
) -> None:
    payload = {"channel": channel, "blocks": blocks, "text": text}
    data = json.dumps(payload).encode()
    req = urllib.request.Request(
        f"{_SLACK_API_BASE}/chat.postMessage",
        data=data,
        headers={
            "Content-Type": "application/json",
            "Authorization": f"Bearer {token}",
        },
        method="POST",
    )
    with urllib.request.urlopen(req, timeout=5) as resp:
        body = json.loads(resp.read())
        if not body.get("ok"):
            raise RuntimeError(f"Slack Web API error: {body.get('error')}")
